src/NFT/AgentCommunication.sol

```diff
@@ -4,19 +4,70 @@ pragma solidity ^0.8.22;
 import {Ownable} from "@openzeppelin/contracts/access/Ownable.sol";
 import "./DoubleEndedStructQueue.sol";
 
-contract AgentCommunication is Ownable {
-    address payable public treasury;
-    uint256 public pctToTreasuryInBasisPoints; //70% becomes 7000
+interface IAgentRegistry {
+    error AgentNotRegistered();
+
+    event AgentRegistered(address indexed agent);
+    event AgentDeregistered(address indexed agent);
+
+    function registerAsAgent() external;
+    function deregisterAsAgent() external;
+    function isRegisteredAgent(address agent) external view returns (bool);
+    function getAllRegisteredAgents() external view returns (address[] memory);
+}
+
+contract AgentRegistry is IAgentRegistry, Ownable {
+    mapping(address => bool) public registeredAgents;
+    address[] private registeredAgentsList;
+
+    constructor() Ownable(msg.sender) {}
+
+    modifier onlyRegisteredAgent() {
+        if (!isRegisteredAgent(msg.sender)) {
+            revert AgentNotRegistered();
+        }
+        _;
+    }
+
+    function registerAsAgent() public {
+        registeredAgents[msg.sender] = true;
+        registeredAgentsList.push(msg.sender);
+        emit AgentRegistered(msg.sender);
+    }
 
-    error MessageNotSentByAgent();
+    function deregisterAsAgent() public onlyRegisteredAgent {
+        registeredAgents[msg.sender] = false;
+        // Remove from list
+        for (uint256 i = 0; i < registeredAgentsList.length; i++) {
+            if (registeredAgentsList[i] == msg.sender) {
+                registeredAgentsList[i] = registeredAgentsList[registeredAgentsList.length - 1];
+                registeredAgentsList.pop();
+                break;
+            }
+        }
+        emit AgentDeregistered(msg.sender);
+    }
+
+    function isRegisteredAgent(address agent) public view returns (bool) {
+        return registeredAgents[agent];
+    }
+
+    function getAllRegisteredAgents() public view returns (address[] memory) {
+        return registeredAgentsList;
+    }
+}
+
+contract AgentCommunication is AgentRegistry {
```

Comment:
Any reason for coupling these 2 contracts?
In my head, it would make more sense that `AgentCommunication` receives an `AgentRegistry` instance (actually the Interface) via constructor arg, and then operate on that.
This way, we keep the contracts independent of each other, since one handles registry and the other handles messaging. Wdty?

Reply:
I tried this approach initially but had some errors with it. I'll try it once more and post them here, maybe you will be able to help. (I like it like that more as well)

Reply:
Cool, happy to huddle or help async.
What I had in mind (hopefully works) is something like this
```
contract AgentRegistry {
[...]
}
contract AgentCommunication is Ownable {

IAgentRegistry public agentRegistry;

constructor(IAgentRegistry agentRegistry) Ownable(msg.sender) {
    agentRegistry = agentRegistry
}

// setter - onlyOwner

}
```
see also https://github.com/seer-pm/demo/blob/4310c1e2b995f675b9a8ef11f9a4f79ed044ecc4/contracts/src/trade/LiquidityManager.sol#L16

Reply:
Seems correct now! We can circle back to it once you are back if something would be off